### PR TITLE
Upgrade to lucene-5.2-snapshot-1675363.

### LINKED ISSUE #10728: Search: FielddataTermsFilter equality is based on hash codes
FielddataTermsFilter considers that two filters are equal if they apply to the same field name and the **hash code** of the wrapped terms are the same. It should compare terms directly.

### PR DESCRIPTION
**Pull request by jpountz**
This snapshot contains in particular LUCENE-6446 (refactored explanation API)
and LUCENE-6448 (better equals/hashcode for filters).

Closes #10728

### CODE REVIEW

File: src/main/java/org/elasticsearch/common/lucene/search/function/CombineFunction.java [hunk 1]
```diff
@@ -35,16 +34,15 @@ public String getName() {
...
-            ComplexExplanation minExpl = new ComplexExplanation(true, Math.min(funcExpl.getValue(), maxBoost), "Math.min of");
-            minExpl.addDetail(funcExpl);
-            minExpl.addDetail(new Explanation(maxBoost, "maxBoost"));
-            res.addDetail(minExpl);
-            res.addDetail(new Explanation(queryBoost, "queryBoost"));
-            return res;
+            Explanation boostExpl = Explanation.match(maxBoost, "maxBoost");
+            Explanation minExpl = Explanation.match(
+                    Math.min(funcExpl.getValue(), maxBoost),
+                    "Math.min of",
```
Comment (rmuir) [inline src/main/java/org/elasticsearch/common/lucene/search/function/CombineFunction.java]: should we just say min of or minimum of?

File: src/main/java/org/elasticsearch/common/lucene/search/function/CombineFunction.java [hunk 2]
```diff
@@ -59,15 +57,15 @@ public String getName() {
...
```
Comment (rmuir) [inline src/main/java/org/elasticsearch/common/lucene/search/function/CombineFunction.java]: same comment about Math.min as before.

File: src/main/java/org/elasticsearch/index/search/FieldDataTermsFilter.java [hunk 3]
```diff
@@ -86,16 +87,19 @@ public static FieldDataTermsFilter newDoubles(IndexNumericFieldData fieldData, D
     @Override
     public boolean equals(Object obj) {
         if (this == obj) return true;
-        if (obj == null || !(obj instanceof FieldDataTermsFilter)) return false;
+        if (super.equals(obj) == false) return false;
```
Comment (rmuir) [inline src/main/java/org/elasticsearch/index/search/FieldDataTermsFilter.java]: wonderful :)

**Comment by rmuir** [comment]
Looks great. I think in general i prefer "min of" or "minimum of" where we currently say "Math.min of", because that only makes sense to java developers. Otherwise I am happy the equals/hashcode are all using superclass methods, and ComplexExplanation is out of here.

**Comment by jpountz** [comment]
Thanks Robert. I kept the explanation descriptions as they were but I don't mind applying the fixes you suggest. We might find more bugs in equals/hashcode as we migrate to queries which recursively call QueryUtils.check on sub queries on the contrary to filters.

**Comment by rmuir** [comment]
Indeed, they were in the CombineFunction before that way (I missed that). I thought they were newly introduced here.

**Pull request merged**